modules/sdk-core/src/bitgo/tss/common.ts

```diff
@@ -68,3 +70,51 @@ export async function sendSignatureShare(
     })
     .result();
 }
+
+/**
+ * Verifies that a TSS wallet signature was produced with the expected key and that the signed data contains the
+ * expected common keychain as well as the expected user and backup key ids
+ */
+export async function commonVerifyWalletSignature(params: {
+  walletSignature: openpgp.Key;
+  bitgoPub: openpgp.Key;
+  commonKeychain: string;
+  userKeyId: string;
+  backupKeyId: string;
+}): Promise<{ value: ArrayBuffer }[]> {
+  const { walletSignature, bitgoPub, commonKeychain, userKeyId, backupKeyId } = params;
+
+  // The result of verifyPrimaryUser contains two items:
+  // * the self-cert of the CA key
+  // * the cert of the subkey (our proof that we are looking for)
+  // By comparing the fingerprints of the walletSignature and the passed bitgoPub we ensure that the passed bitgoPub is
+  // not the CA key (if they are the same, then it is).
+  if (walletSignature.keyPacket.getFingerprint() === bitgoPub.keyPacket.getFingerprint()) {
```

Comment:
i really don't see the need for this.   especially after verifying the behavior of the `.verifyPrimaryUser` function

Reply:
I still do. If I remove this as well but leave the rest of the code the same then the following test will fail:
https://github.com/BitGo/BitGoJS/blob/79da30b69fb7be7a87d27f2ca463e751c20dda04/modules/bitgo/test/v2/unit/tss/common.ts#L38-L47
What this test is really doing is that a man in the middle is using their own gpg key (key1) to produce a proof over the BitGo pub (key2). This test should fail and it only fails if I keep this line.

Reply:
that's not a MitM attack, that's not using the SDK properly.

the point of `commonVerifyWalletSignature()` is to take a pubkey, such as walletSignature, and verify if the second argument (bitgoPub in this case) certifies that same pubkey.     

in your test, you have key1 certifying key2 (you produce a key2 pub with a cert attached from key1).    then you pass in this certification (by passing in key2 pub with any certs attached), and bitgoPub/certifying key set to key2.   You're essentially asking the function to make sure that key2 has a certifcation from....key2 (and not key1, like you would expect).    Considering that bitgoPub is baked into the SDK, and should always be passed into the function, the only way to abuse this would be to modify this code (which wouldn't be mitm, which happens with the communication between parties).

I....guess I can agree it's good to keep guard rails here.   but if you're that concerned with someone using this code incorrectly, then i would argue that passing in bitgoPub at all should be removed and instead use the global defined bitgo public key.

Reply:
I wasn't aware that bitgoPub is baked into the SDK. So far I only saw code that pulls it once from BitGo's `/api/v1/client/constants` API and then caches it.

Reply:
Where can I find the baked in bitgoPub?

Reply:
wow, I thought it was.   I guess that is only OVC and on platform.

Then yeah, we *should* bake this into SDK then.   that would help solve MitM problems.

Reply:
testnet: https://github.com/BitGo/bitgo-microservices/blob/56dd0f77c6a9878c8c55798ad1b903d6ecba4753/packages/wallet-platform/config/env/testnet.js#L145-L163

production:  https://github.com/BitGo/bitgo-microservices/blob/56dd0f77c6a9878c8c55798ad1b903d6ecba4753/packages/wallet-platform/config/env/production.js#L241-L254

I never liked the idea of pulling these from a remote API, since we should have a trust anchor in the SDK itself.

Reply:
I think we should look at this function in isolation and think of what it is promising, which is a robust check that walletSig has a non-self certification that is signed by bitgoPub and that the certification contains notation data that matches the wallet data that all parties have received so far in the creation of the wallet. Therefore, although baking bitgoPub would be the safest thing to do, the code as is will be passing it as a parameter, so I believe this check should stay there, otherwise if at some point bitgo's Pub key is baked as a global variable, I agree then that the check would be useless.

Reply:
But also, I just realized this check protects against the case if a MiM intercepts the communication between WP and User,  then replaces walletSig with their own walletSig that is just HSM's public key without even any other certs attached to it. In that case, wouldn't checking walletSig using bitgoPub through verifyPrimaryUser return valid without the above check? Because walletSig.verifyPrimaryUser([bitgoPub]) would return valid being true for the self certification in walletSig since the attacker just passed bitgoPub in the communication.

Reply:
@islamaminBitGo well I'm making the argument that we SHOULD bake it in.  It's important to have a trust anchor on the SDK itself, as MitM attacks could occur (this is why ovc even includes it)

> But also, I just realized this check protects against the case if a MiM intercepts the communication between WP and User, then replaces walletSig with their own walletSig that is just HSM's public key without even any other certs attached to it.

But the whole point of walletsigs, is that you get two GPG pubkeys and (after verifying the HSM has blessed them with a cert), you then cross-check that those are the gpg keys you've already learned are the two gpg keys to use (one being yourself, if you are the user, and the other being the backup).   Not only do you check that the HSM has certified those two keys, but that the certification itself includes notation data that specifies the two keyID's AND the commonPub (and also u-values).    

So if an attacker swapped out one of the walletsigs / user pubkey with the HSM one, then you should _still_ have a failure because the keyIDs are not lining up (you already know the gpg key of yourself, and you already know the gpg key of the backup provider) so you are able to check.   So if this single `if` statement is making that big of a difference, then we have evidence that our checking of these other things are not being performed correctly.

Reply:
Ah, exactly so one thing we're missing here in this check is an initial check of whether the two keys in walletSigs are indeed user's and backup's. 

Reiterating here to see if this is one way to move forward: 
Add one check in the function that dissects walletSigs into userGpgKeyWithCertAdded and backupGpgKeyWithCertAdded:
(Assume we're user and do the same thing for backup)
Check:
userGpgKeyWithCertAdded has the same fingerprint as our own key's fingerprint
backupGpgKeyWithCertAdded has the same fingerprint as that of the pub key we hold for backup

Reply:
yup, you got it.

the commonKeychain should also be checked too, but I guess that's done in the other function.